examples/pytorch/gcmc/gcmc_mp_sampling.py

```diff
@@ -0,0 +1,470 @@
+            self.encoder.to(dev_id)
+
+        self.decoder = SampleBiDecoder(args.rating_vals,
+                                       in_units=args.gcn_out_units,
+                                       num_basis_functions=args.gen_r_num_basis_func)
+        self.decoder.to(dev_id)
+
+    def forward(self, compact_g, frontier, ufeat, ifeat, possible_rating_values):
+        user_out, movie_out = self.encoder(frontier, frontier, ufeat, ifeat)
+
+        head_emb = []
+        tail_emb = []
+        for possible_rating_value in possible_rating_values:
+            head, tail = compact_g.all_edges(etype=str(possible_rating_value))
+            head_emb.append(user_out[head])
+            tail_emb.append(movie_out[tail])
+
+        head_emb = th.cat(head_emb, dim=0)
+        tail_emb = th.cat(tail_emb, dim=0)
+
+        pred_ratings = self.decoder(head_emb, tail_emb)
+        return pred_ratings
+
+def evaluate(args, dev_id, net, dataset, segment='valid'):
+    possible_rating_values = dataset.possible_rating_values
+    nd_possible_rating_values = th.FloatTensor(possible_rating_values).to(dev_id)
+
+    if segment == "valid":
+        rating_values = dataset.valid_truths
+        enc_graph = dataset.valid_enc_graph
+        dec_graph = dataset.valid_dec_graph
+    elif segment == "test":
+        rating_values = dataset.test_truths
+        enc_graph = dataset.test_enc_graph
+        dec_graph = dataset.test_dec_graph
+    else:
+        raise NotImplementedError
+
+    num_edges = rating_values.shape[0]
+    seed = th.arange(num_edges)
+    count_rmse = 0
+    count_num = 0
+    real_pred_ratings = []
+    true_rel_ratings = []
+    for sample_idx in range(0, (num_edges + args.minibatch_size - 1) // args.minibatch_size):
+        net.eval()
+        start = sample_idx * args.minibatch_size
+        end = (sample_idx + 1) * args.minibatch_size \
+              if (sample_idx + 1) * args.minibatch_size < num_edges else num_edges
+        edge_ids = seed[start:end]
+
+        sub_rating_values = rating_values[edge_ids]
+        head_id, tail_id = dec_graph.find_edges(edge_ids)
+        utype, _, vtype = enc_graph.canonical_etypes[0]
+        
+        subg = []
+        sub_true_rel_ratings = []
+        for possible_rating_value in possible_rating_values:
+            idx_loc = (sub_rating_values == possible_rating_value)
+            head = head_id[idx_loc]
+            tail = tail_id[idx_loc]
+            sub_true_rel_ratings.append(sub_rating_values[idx_loc])
+            subg.append(dgl.bipartite((head, tail),
+                                        utype=utype,
+                                        etype=str(possible_rating_value),
+                                        vtype=vtype,
+                                        card=(enc_graph.number_of_nodes(utype),
+                                              enc_graph.number_of_nodes(vtype))))
+
+        g = dgl.hetero_from_relations(subg)
+        g = dgl.compact_graphs(g)
```

Comment:
Reuse the sampler code?

Reply:
Yes. For eval, we also can use dataloader.